### PR TITLE
msg/async: Skip the duplicated processing of the same link

### PR DESCRIPTION
**Pull request by shangfufei**
msg/async: Skip the processing of the same link, when the external processing function is called. By doing this, the efficiency of event processing can be improved.

Signed-off-by: shangfufei <shangfufei@inspur.com>

### CODE REVIEW

File: src/msg/async/Event.cc [hunk 1]
```diff
@@ -428,13 +428,19 @@ int EventCenter::process_events(unsigned timeout_microseconds,  ceph::timespan *
...
     deque<EventCallbackRef> cur_process;
+    EventCallbackRef prev_call_back = NULL;
     cur_process.swap(external_events);
     external_num_events.store(0);
     external_lock.unlock();
     numevents += cur_process.size();
     while (!cur_process.empty()) {
       EventCallbackRef e = cur_process.front();
       ldout(cct, 30) << __func__ << " do " << e << dendl;
+      if (prev_call_back == e){
```
Comment (tchaikov) [inline src/msg/async/Event.cc]: can we do the dedup in `EventCenter::dispatch_event_external()`? assuming that the worker thread will always be able to pickup the consecutive same events pending in the queue in batch.

Comment (shangfufei) [reply src/msg/async/Event.cc]: This situation should be in a scenario where a worker thread has more than one connection and there is only one read-write event at the same time, especially under the pressure of large-scale cluster.

Even with this extreme situation, it can be evaded by configuration switches.

Comment (tchaikov) [reply src/msg/async/Event.cc]: @shangfufei sorry, i don't understand your reply. what i suggest is to avoid appending new callbacks to `external_events`, if it already has the same callback at its end. as i think that's the case addressed by your fix.

Comment (shangfufei) [reply src/msg/async/Event.cc]: void EventCenter::dispatch_event_external(EventCallbackRef e)
{
  external_lock.lock();
  if(NULL == external_events.find(e))
  {
    external_lock.unlock();
    return;
  }
  external_events.push_back(e);
  bool wake = !external_num_events.load();
  uint64_t num = ++external_num_events;
  external_lock.unlock();
  if (!in_thread() && wake)
    wakeup();

  ldout(cct, 20) << __func__ << " " << e << " pending " << num << dendl;
}
At first I changed this way. but i think he search efficiency of std:: deque is relatively low, which may affect the efficiency of the program.

Comment (shangfufei) [reply src/msg/async/Event.cc]: This is  an example of pseudo code.

Comment (tchaikov) [reply src/msg/async/Event.cc]: i don't think your pseudo code is equivalent to what you have in this change, instead i'd suggest something like:
```c++
void EventCenter::dispatch_event_external(EventCallbackRef e)
{
  std::lock_guard lock{external_lock};
  bool was_empty = !external_num_events.load();
  if (!was_empty && *external_events.rbegin() == e) {
    return;
  }
  external_events.push_back(e);
  uint64_t num = ++external_num_events;
  if (was_empty && !in_thread()) {
    wakeup();
  }
  ldout(cct, 30) << __func__ << " " << e << " pending " << num << dendl;
}
```

please note, in your change you are always comparing the n and n+1 callback. instead doing the global dedup.

but after a second thought, i wonder, if the thread which

Comment (shangfufei) [reply src/msg/async/Event.cc]: I see that the code you modify has the following questions:

1. the scope of lock external_lock will increase. I am not sure whether it will cause other problems.

2. The way the code changes its way of thinking is really improving with my previous changes. The following steps can be omitted: do not put duplicates in the queue, then take them out and release them.

Comment (shangfufei) [reply src/msg/async/Event.cc]: void EventCenter::dispatch_event_external(EventCallbackRef e)
{
  external_lock.lock();
  bool was_empty = !external_num_events.load();
  if (!was_empty && *external_events.rbegin() == e) {
  	external_lock.unlock();  
    return;
  }
  external_events.push_back(e);
  uint64_t num = ++external_num_events;
  external_lock.unlock();
  if (was_empty && !in_thread()) {
    wakeup();
  }
  ldout(cct, 30) << __func__ << " " << e << " pending " << num << dendl;
}

Will this be more insurance?

Comment (tchaikov) [reply src/msg/async/Event.cc]: > The way the code changes its way of thinking is really improving with my previous changes. The following steps can be omitted: do not put duplicates in the queue, then take them out and release them.

could you rephrase this? and why is this a "problem"?

> the scope of lock external_lock will increase. I am not sure whether it will cause other problems.

if we want to limit the scope of lock, probably we can:

```c++
void EventCenter::dispatch_event_external(EventCallbackRef e)
{
  uint64_t num = 0;
  {
    std::lock_guard lock{external_lock};
    if (external_num_events > 0 && *external_events.rbegin() == e) {
      return;
    }
    external_events.push_back(e);
    num = ++external_num_events;
  }
  if (num == 1 && !in_thread()) {
    wakeup();
  }
  ldout(cct, 30) << __func__ << " " << e << " pending " << num << dendl;
}
```

File: src/msg/async/Event.cc [hunk 2]
```diff
@@ -452,12 +452,16 @@ int EventCenter::process_events(unsigned timeout_microseconds,  ceph::timespan *
...
+  uint64_t num = 0;
+  {
+    std::lock_guard lock{external_lock};
+    if (external_num_events > 0 && *external_events.rbegin() == e) {
+      return;
+    }
+    external_events.push_back(e);
+    num = ++external_num_events;
+  }
+  if (!in_thread() && 1 == num)
```
Comment (tchaikov) [inline src/msg/async/Event.cc]: i'd suggest put
```c++
if (num == 1 && !in_thread()) {
```
for better performance and to be more consistent. yoda conditions are not forbidden but i failed to find its occurrences in https://google.github.io/styleguide/cppguide.html#Conditionals.

**Comment by tchaikov** [comment]
please fold these two commit into a single one and could you add a "`Signed-off-by`" line at the end of your commit message? "git commit -s" will do the trick for you. see https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#1-sign-your-work

File: src/msg/async/Event.cc (hunk 2 above)
Comment (shangfufei) [reply src/msg/async/Event.cc]: ok, that's better, thx for your remind.

**Comment by liewegas** [comment]
@shangfufei can you squash this into a single commit please?

**Comment by shangfufei** [comment]
> @shangfufei can you squash this into a single commit please?

I have received it. Thank you for reminding me. I will deal with it as soon as possible.

**Comment by tchaikov** [comment]
@shangfufei you might need to fold them into a *single* commit.

**Comment by shangfufei** [comment]
I have made a merger of the first three submissions.

**Comment by tchaikov** [comment]
and could you squash 358fa87bbed08eb61efddbf8fb06d0153be05d3a into b961b35b7c10ead183cd49cd23d6ca315aec732b ?

**Comment by shangfufei** [comment]
I did it, but the annotation also followed the merger. Next time, it is necessary to unify the annotation.

**Comment by tchaikov** [comment]
@shangfufei i can still see two commits in your PR.  see:
```
$ git log wip-20952 --oneline
358fa87bbe (wip-20952) Update Event.cc
b961b35b7c msg/async: Skip the processing of the same link, when the external processing function is called.
9b46b9723b Merge pull request #20939 from dillaman/wip-cls-rbd-group
```
i am not sure what "annotation" stands for.

**Comment by tchaikov** [comment]
@shangfufei could you update the title of the commit message? so it looks like
```
msg/async: Skip the duplicated processing of the same link
```

**Comment by shangfufei** [comment]
No problem. It's changed.

**Comment by tchaikov** [comment]
@shangfufei you need to update the commit and *repush*. please note, the title of PR is different from that of commit.

**Comment by tchaikov** [comment]
@shangfufei currently the title of your commit is "Update Event.cc", which is not as helpful as "msg/async: Skip the duplicated processing of the same link." and ideally, you could add an empty line before your Signed-off-by line.

**Comment by shangfufei** [comment]
Thank you very much for your suggestion. is that ok ?

**Comment by tchaikov** [comment]
- http://pulpito.ceph.com/kchai-2018-11-29_01:22:06-rados-wip-kefu2-testing-2018-11-29-0053-distro-basic-smithi/
- http://pulpito.ceph.com/kchai-2018-11-29_08:14:41-rados-wip-kefu2-testing-2018-11-29-1356-distro-basic-smithi/

test failures are tracked / caused by

- #19454
- http://tracker.ceph.com/issues/37391
